### PR TITLE
12.0 IMP sale_order_lot_selection: allow to confirm sale when lot unavailable

### PR DESCRIPTION
**Pull request by bealdav**
You may want use unavailable lot (i.e. you sell a product manufactured by a supplier
but with your requirements. Then you provide a lot number before to receive product

### CODE REVIEW

**Comment by rvalyi** [comment]
I understand perfectly you may want to sell new lots (not available then) and this is a great improvement.

But then that would be an assumed CHOICE. With your PR what about a company who would still want to avoid selling unavailable lots in its own use case? I mean I feel there should be an option in the the company, settings or sale team and leave the former `_check_move_state()` as an option, no?

**Comment by bealdav** [comment]
You're right, I'll put an overidable method to allow to skip

**Comment by bealdav** [comment]
Fail comes from 12.0 sale_stock_secondary_unit and sale_elaboration

File: sale_order_lot_selection/models/sale.py [hunk 1]
```diff
@@ -52,32 +52,28 @@ def get_move_from_line(self, line):
...
-                move.refresh()
-            if move.state != 'assigned':
-                raise UserError(_('Can\'t reserve products for lot %s') %
-                                line.lot_id.name)
+    def forbid_unavailable_lot_reservation(self):
+        """ You may want use unavailable lot in your stock move:
+            i.e. you sell a product manufactured by a supplier
+            but with your requirements including traceability on received lot.
+
+            In this case override this method to return True in your use case
```
Comment (rvalyi) [inline sale_order_lot_selection/models/sale.py]:
```suggestion
            In this case override this method to return False in your use case
```

File: sale_order_lot_selection/models/sale.py [hunk 2]
```diff
@@ -52,32 +52,28 @@ def get_move_from_line(self, line):
...
-                    move.reserved_availability
-                )
-                if unreserved_moves:
+                move = self.get_move_from_line(line)
+                if move.state == 'confirmed':
+                    move._action_assign()
+                    move.refresh()
+                if (
+                    move.state != "assigned"
+                    and self.forbid_unavailable_lot_reservation()
```
Comment (rvalyi) [inline sale_order_lot_selection/models/sale.py]: would be better to pass the line and/or the move to this forbid_unavailable_lot_reservation hook (so we can have the lot and the product), so you can do the necessary checking that you are indeed in your special case, no?

**Review by rvalyi — DISMISSED** [review]

File: sale_order_lot_selection/models/sale.py (hunk 1 above)
Comment (rvalyi) [reply sale_order_lot_selection/models/sale.py]: (True is the default here)

**Comment by bealdav** [comment]
Concerned by https://github.com/OCA/sale-workflow/issues/1166

**Comment by bealdav** [comment]
Hi @eLBati @PCatinean I've made changes explained https://github.com/OCA/sale-workflow/issues/1166

Could you make a review please ?

**Comment by PCatinean** [comment]
Hi @bealdav, either it's been ages since I've been working with Gitlab only and forgot how to use Github or something else is wrong. The diff shows just removal of functionality from sale.py, some demo data addition and changes in tests. Am I missing something?

**Comment by bealdav** [comment]
You're right the diff show removing code as explained in #1166 
Because this code prevent a behaviour (you can't confirm a sale order when no lot is not available)

**Comment by PCatinean** [comment]
Ah understood, I was referring to the conversation you had with rvalyi and was looking for the optional method you mentioned. Will you still be adding that?

**Comment by bealdav** [comment]
Yes but developers (you and @eLBati) can't remember (even searching in their history) why this code was added. Then here 2 possibilities

- we keep these features for ever even nobody know what or who they are suppose to serve (see https://en.wikipedia.org/wiki/The_Cathedral_and_the_Bazaar)

- we clean undocumented features (problably needed when reach an older version for a specific project) with updated code matching with normal behavior of odoo at this version

If we want to keep OCA great for the future I think it's better to move towards the second strategy

**Review by PCatinean — APPROVED** [review]

**Review by sebastienbeau — APPROVED** [review]
LGTM, ok for dropping unused feature. In any case if we want to add teh contstrait it should be done in a separated module

**Comment by sebastienbeau** [comment]
@rvalyi can you finish you review?

**Review by marcelsavegnago — APPROVED** [review]

**Comment by sebastienbeau** [comment]
@rousseldenis Can you merge it ? (I do not have the right anymore)

**Comment by bealdav** [comment]
Please @rousseldenis could you merge it ? Thank you.

**Comment by bealdav** [comment]
ping @rousseldenis  could you have a look ?

**Pull request merged**